### PR TITLE
Better support for dup syscall family

### PR DESCRIPTION
**Pull request by alacuku**
```
 Signed-off-by: Aldo Lacuku <aldo@lacuku.eu>
```

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area driver-kmod

/area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-udig

/area libscap

/area libsinsp

> /area tests

> /area proposals

**What this PR does / why we need it**:
The current implementation sends to userspace the same type of event each time one of the following sycalls is called:
```
int dup(int oldfd);
int dup2(int oldfd, int newfd);
int dup3(int oldfd, int newfd, int flags);
```
Meaning that the calls to `dup2` and `dup3` are masqueraded as `dup` events. 

This PR implements each syscall with its own type of event and number of parameters. For more info see the message of each commit.
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
update(driver): better support for dup syscall family (dup, dup2, and dup3)
```

### CODE REVIEW

File: userspace/libsinsp/parsers.cpp [hunk 1]
```diff
@@ -408,6 +408,12 @@ void sinsp_parser::process_event(sinsp_evt *evt)
 	case PPME_SYSCALL_DUP_X:
 		parse_dup_exit(evt);
 		break;
+	case PPME_SYSCALL_DUP2_X:
+		parse_dup_exit(evt);
+		break;
+	case PPME_SYSCALL_DUP3_X:
+		parse_dup_exit(evt);
```
Comment (Andreagit97) [inline userspace/libsinsp/parsers.cpp:415]: Maybe we can get also the flag information for `dup3` in the `sinsp_fdinfo_t`. WDYT about it?

**Review by Andreagit97 — COMMENTED** [review]
Thank you @alacuku great job!

File: driver/event_table.c [hunk 2]
```diff
@@ -136,8 +136,8 @@ const struct ppm_event_info g_event_info[PPM_EVENT_MAX] = {
 	/* PPME_SYSCALL_PREADV_X */{"preadv", EC_IO_READ, EF_USES_FD | EF_READS_FROM_FD | EF_DROP_SIMPLE_CONS, 3, {{"res", PT_ERRNO, PF_DEC}, {"size", PT_UINT32, PF_DEC}, {"data", PT_BYTEBUF, PF_NA} } },
 	/* PPME_SYSCALL_PWRITEV_E */{"pwritev", EC_IO_WRITE, EF_USES_FD | EF_WRITES_TO_FD | EF_DROP_SIMPLE_CONS, 3, {{"fd", PT_FD, PF_DEC}, {"size", PT_UINT32, PF_DEC}, {"pos", PT_UINT64, PF_DEC} } },
 	/* PPME_SYSCALL_PWRITEV_X */{"pwritev", EC_IO_WRITE, EF_USES_FD | EF_WRITES_TO_FD | EF_DROP_SIMPLE_CONS, 2, {{"res", PT_ERRNO, PF_DEC}, {"data", PT_BYTEBUF, PF_NA} } },
-	/* PPME_SYSCALL_DUP_E */{"dup", EC_IO_OTHER, EF_CREATES_FD | EF_USES_FD | EF_MODIFIES_STATE, 1, {{"fd", PT_FD, PF_DEC} } },
-	/* PPME_SYSCALL_DUP_X */{"dup", EC_IO_OTHER, EF_CREATES_FD | EF_USES_FD | EF_MODIFIES_STATE, 1, {{"res", PT_FD, PF_DEC} } },
+	/* PPME_SYSCALL_DUP_E */{"dup", EC_IO_OTHER, EF_NONE, 0 },
```
[the remark this answers is not in the document]
Comment (FedeDP) [reply driver/event_table.c:139]: I agree with Andrea; i'd rather:
* leave the enter event as is, and only update the exit event
* add entirely new PPME_SYSCALL_DUP_{E,X}_1

Rule of thumb is: if you add params, it's ok. If you remove params, ko.

IMO PPME_SYSCALL_DUP_{E,X}_1 is a good choice :)

File: driver/event_table.c [hunk 3]
```diff
@@ -374,6 +374,10 @@ const struct ppm_event_info g_event_info[PPM_EVENT_MAX] = {
 	/* PPME_GROUP_ADDED_X */{"groupadded", EC_PROCESS, EF_UNUSED, 0},
 	/* PPME_GROUP_DELETED_E */{"groupdeleted", EC_PROCESS, EF_MODIFIES_STATE, 3, {{"gid", PT_UINT32, PF_DEC}, {"name", PT_CHARBUF, PF_NA}, {"container_id", PT_CHARBUF, PF_NA} } },
 	/* PPME_GROUP_DELETED_X */{"groupdeleted", EC_PROCESS, EF_UNUSED, 0},
+	/* PPME_SYSCALL_DUP2_E */{"dup2", EC_IO_OTHER, EF_NONE, 0},
+	/* PPME_SYSCALL_DUP2_X */{"dup2", EC_IO_OTHER, EF_CREATES_FD | EF_USES_FD | EF_MODIFIES_STATE, 3, {{"res", PT_FD, PF_DEC}, {"oldfd", PT_FD, PF_DEC}, {"newfd", PT_FD, PF_DEC} } },
+	/* PPME_SYSCALL_DUP3_E */{"dup3", EC_IO_OTHER, EF_NONE, 0},
+	/* PPME_SYSCALL_DUP3_X */{"dup3", EC_IO_OTHER, EF_CREATES_FD | EF_USES_FD | EF_MODIFIES_STATE, 4, {{"res", PT_FD, PF_DEC}, {"oldfd", PT_FD, PF_DEC}, {"newfd", PT_FD, PF_DEC}, {"flags", PT_UINT64, PF_HEX, dup3_flags} } },
```
[the remark this answers is not in the document]
Comment (alacuku) [reply driver/event_table.c]: Nice catch, thanks!

File: userspace/libsinsp/parsers.cpp (hunk 1 above)
Comment (alacuku) [reply userspace/libsinsp/parsers.cpp:415]: Not sure if it is useful. @FedeDP what do you think?

Comment (FedeDP) [reply userspace/libsinsp/parsers.cpp:415]: I am not sure either; we are only setting the O_CLOEXEC flag right?
It seems that fdinfo has only these flags:
```
enum flags
	{
		FLAGS_NONE = 0,
		FLAGS_FROM_PROC = (1 << 0),
		//FLAGS_TRANSACTION = (1 << 1),
		FLAGS_ROLE_CLIENT = (1 << 2),
		FLAGS_ROLE_SERVER = (1 << 3),
		FLAGS_CLOSE_IN_PROGRESS = (1 << 4),
		FLAGS_CLOSE_CANCELED = (1 << 5),
		FLAGS_IS_SOCKET_PIPE = (1 << 6),
		FLAGS_IS_TRACER_FILE = (1 << 7),
		FLAGS_IS_TRACER_FD = (1 << 8),
		FLAGS_IS_NOT_TRACER_FD = (1 << 9),
		FLAGS_IN_BASELINE_R = (1 << 10),
		FLAGS_IN_BASELINE_RW = (1 << 11),
		FLAGS_IN_BASELINE_OTHER = (1 << 12),
		FLAGS_SOCKET_CONNECTED = (1 << 13),
		FLAGS_IS_CLONED = (1 << 14),
		FLAGS_CONNECTION_PENDING = (1 << 15),
		FLAGS_CONNECTION_FAILED = (1 << 16),
	};
```

@Andreagit97 am i right?

Comment (Andreagit97) [reply userspace/libsinsp/parsers.cpp:415]: This is right, but I was thinking about [`m_openflags` field](https://github.com/falcosecurity/libs/blob/78f67c4b64cd7ea3d407facfa169de953fd10750/userspace/libsinsp/fdinfo.h#L336) where we store information about open flags and `dup3` uses one of this open flags `O_CLOEXEC`. This is similar to what we do [here](https://github.com/falcosecurity/libs/blob/78f67c4b64cd7ea3d407facfa169de953fd10750/userspace/libsinsp/parsers.cpp#L2409). WDYT @alacuku @FedeDP?

Comment (FedeDP) [reply userspace/libsinsp/parsers.cpp:415]: +1 from me, thanks Andre for spotting it!

**Comment by leogr** [comment]
Hey @alacuku 

Once this gets merged and Falco bumped with a new driver version, please take care of :point_down: 
https://github.com/falcosecurity/falco/blob/85f91a3ec44b9cd81f8d678f0375fe77194b4e00/rules/falco_rules.yaml#L2986-L2991

File: userspace/libsinsp/parsers.cpp [hunk 4]
```diff
@@ -4395,9 +4397,14 @@ void sinsp_parser::parse_dup_exit(sinsp_evt *evt)
...
-
 		//
-		// If the old FD is in the table, remove it properly
+		// If the old FD is in the table, remove it properly.
+		// The old FD is:
+		// 	- dup(): fd number of a previously closed fd that has not been removed from the fd_table
+		//		     and has been reassigned to the newly created fd by dup()(very rare condition);
+		//  - dup2(): fd number of an existing fd that we pass to the dup2() as the "newfd". dup2()
+	    //			  will close the existing one. So we need to clean it up / overwrite;
+		//  - dup3(): same as dup2().
```
Comment (FedeDP) [inline userspace/libsinsp/parsers.cpp:4407]: Thanks for these comments <3

**Review by FedeDP — COMMENTED** [review]
Changes LGTM! Moreover, the digging work here is remarkable, thanks for the great job!

Left a small comment!

**Comment by alacuku** [comment]
> Hey @alacuku
> 
> Once this gets merged and Falco bumped with a new driver version, please take care of point_down https://github.com/falcosecurity/falco/blob/85f91a3ec44b9cd81f8d678f0375fe77194b4e00/rules/falco_rules.yaml#L2986-L2991

@leogr, for sure! Thanks!

File: driver/event_table.c [hunk 5]
```diff
@@ -374,6 +374,12 @@ const struct ppm_event_info g_event_info[PPM_EVENT_MAX] = {
 	/* PPME_GROUP_ADDED_X */{"groupadded", EC_PROCESS, EF_UNUSED, 0},
 	/* PPME_GROUP_DELETED_E */{"groupdeleted", EC_PROCESS, EF_MODIFIES_STATE, 3, {{"gid", PT_UINT32, PF_DEC}, {"name", PT_CHARBUF, PF_NA}, {"container_id", PT_CHARBUF, PF_NA} } },
 	/* PPME_GROUP_DELETED_X */{"groupdeleted", EC_PROCESS, EF_UNUSED, 0},
+	/* PPME_SYSCALL_DUP2_E */{"dup2", EC_IO_OTHER, EF_CREATES_FD | EF_USES_FD | EF_MODIFIES_STATE, 1, {{"fd", PT_FD, PF_DEC} } },
```
[the remark this answers is not in the document]
Comment (alacuku) [reply driver/event_table.c:377]: I added a few lines to explain the logic behind  `EF_USES_FD`.

**Review by FedeDP — APPROVED** [review]
/approve

**Pull request merged**